dynamiqs/solvers/propagator.py

```diff
@@ -28,20 +29,38 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
         # check that Hamiltonian is time-independent
-        if not isinstance(self.H, ConstantTimeTensor):
+        if not isinstance(self.H, (ConstantTimeTensor, PWCTimeTensor)):
             raise TypeError(
-                'Solver `Propagator` requires a time-independent Hamiltonian.'
+                f'Solver {obj_type_str(self.options.solver)} requires a'
+                ' time-independent or piecewise-constant Hamiltonian (types'
+                ' `ArrayLike`, `ConstantTimeTensor` or `PWCTimeTensor`), but the'
+                f' passed Hamiltonian has type {obj_type_str(self.H)}.'
             )
-        self.H = self.H(self.t0)
+        if isinstance(self.H, PWCTimeTensor):
+            # We build a custom times array `tstop_new` gathering `self.tstop` and
+            # `self.H.times`. This ensures that each PWC part of the Hamiltonian is
+            # taken into account during the evolution.
+
+            # keep only `H.times` values in the interval [t0, tsave[-1])
+            mask = (self.t0 <= self.H.times) & (self.H.times < self.tsave[-1])
+            extra_tstop = self.H.times[mask]
+            extra_tstop = extra_tstop.numpy(force=True)
+
+            # build new times to stop
+            self.tstop_new = np.unique(np.concatenate((self.tstop, extra_tstop)))
+        else:
+            self.tstop_new = self.tstop
 
     def run_autograd(self):
         t1, y = self.t0, self.y0
-        for t2 in tqdm(self.tstop, disable=not self.options.verbose):
+        for t2 in tqdm(self.tstop_new, disable=not self.options.verbose):
             if t2 != self.t0:
                 # round time difference to avoid numerical errors when comparing floats
                 delta_t = round_truncate(t2 - t1)
                 y = self.forward(t1, delta_t, y)
-            self.save(y)
+            # if the time is in the original `self.tstop` times, save the result
+            if t2 in self.tstop:
```

Comment:
And here I guess this should be `self.tsave` instead? i.e. we stop on `self.tstop` and we save on `self.tsave`. Makes more sense I think.

Reply:
This goes a bit against the base class `tstop` logic. `tstop` = the base class asks to stop (and save or does whatever it needs to do), `tstop_new` = the propagator subclass asks to stop. I think it might be harmful to use `self.tsave` here, it breaks the base class abstraction.

Reply:
Actually why do you need to check `if t2 in self.tstop` ? This check should already be achieved within `self.save` with the `self.tsave_mask`. Otherwise maybe we should adapt it there?

Reply:
I'm not sure to understand. If you read the first bullet point of previous comment, this is how the base solver class was designed: subclasses must stop at every time in `tstop`, and call `save()`. I think it's the clearest way to manage saving indepently of what the solver does.